### DISCUSSION TITLE
Protect .yml file with encryption

### DISCUSSION THREAD

**Discussion by fnfontana** [espanso/espanso#1270] [Feature requests and Ideas]
**Is your feature request related to a problem? Please describe.**
The possibility to protect sensitive data by encrypting the YAML file.
On a daily basis, I've felt the necessity to store personal information, like document ID's, passport, personal address, email, etc.
But I don't think it's safe for us to store sensitive information without be encrypted.

**Describe the solution you'd like**
It would be helpful if Espanso could encrypt this information and make it available by validating the user identity.

**Describe alternatives you've considered**
Entering a password, or PIN, or even use biometrical validation like Windows Hello in the first time we use it on the logon session.

**Additional context**
KeepassXC is a great example, but Espanso don't that all that robust solution, just enough to avoid information leak by someone sharing the same computer or if the .yml file fall on the wrong hands.

**Comment by federico-terzi** [comment]
Hey @fnfontana,

Thank you for the feedback! Espanso was never designed to store sensitive information, so it's unlikely this kind of use-cases will be supported in the near future. 
That said, you could get pretty far by adding your sensitive data to a password manager that has a cli tool (like [Bitwarden CLI](https://bitwarden.com/help/cli/)) and then interface to it using the [shell extension](https://espanso.org/docs/matches/extensions/#shell-extension)

That way, the sensitive data stays is managed by a password manager which then sends it to Espanso only when needed

Cheers :)

**Comment by cgasp** [comment]
@fnfontana , as workaround you can use the system keyring. I've a small python script 'get_keyring_pass' which pull data from keyring. 

```
#!/usr/bin/env python3
import sys
import keyring

# md5sum of service_id
if sys.argv[1] == "429c5f7180e54c228e1b2db67bf3b5b2":
    service_id = "my_ID"
    print(keyring.get_password(service_id, 'username'))
```
more [here](https://stackoverflow.com/a/48831213) to set keyring password.  

Then in espanso config 

```
  - trigger: "!id"
    replace: "{{myshell}}"
    vars:
      - name: myshell
        type: shell
        params:
          cmd: "get_keyring_pass 429c5f7180e54c228e1b2db67bf3b5b2"
```

**Reply by ypurpl** [reply]
I know  this is old, but this helped me a lot

**Comment by fnfontana** [comment]
It's a very cleaver solution, thank you for sharing! I'm currently using keepassxc to do the trick, now I will experiment this solution asap.

**Comment by smeech** [comment]
As there appears to be a number of secure solutions, and Espanso is unlikely to try to duplicate them, I'm inclined to close this Discussion. Let me know if you think it should stay open.
